xdsl/irdl/constraints.py

```diff
@@ -439,6 +439,20 @@ def __init__(
         constrs: tuple[GenericAttrConstraint[AttributeCovT], ...] = tuple(
             irdl_to_attr_constraint(constr) for constr in attr_constrs
         )
+
+        bases = set[Attribute]()
+
+        for c in constrs:
+            b = c.get_bases()
+            if b is None:
+                raise PyRDLError(
+                    f"Constraint {c} cannot appear in an `AnyOf` constraint as its bases aren't known."
+                )
+            if not b.isdisjoint(bases):
+                raise PyRDLError(
+                    f"Constraints {constrs} do not have disjoint bases so cannot be put in an `AnyOf` constraint."
+                )
```

Comment:
So in the process of writing a test I have found I never updated `bases`

Reply:
I don't understand. i32 | i64 seem disjoint to me?

Reply:
They have the same base, which is what this pr uses to judge disjointness. Would be happy to consider a different disjointness criterion

Reply:
Ah, hmm. We could consider special casing `|` on ParamAttrConstraint to fold the `|` into the children if the base is the same.

Reply:
We could consider making eq constraints ParamAttrConstraints :)

Reply:
Would still need to add an bunch more int constraint stuff for this. Doesn't feel like the way to go

Reply:
Can you mark is as draft and push the changes that make things fail so that I can play with them?